### ISSUE TITLE
QR code is not generated with new account

### ISSUE THREAD

**Issue by Akruidenberg** [authelia/authelia#1211]
Today, i´ve setup a new installation with an LDAP, MySQL and redis server. 
Everything is working fine, login with LDAP is working correctly. However, when i choose the option to send and email to generate an QR code, the code is not generated. see the screenshot bellow

![Schermafdruk van 2020-07-18 16-13-16](https://user-images.githubusercontent.com/18529293/87854543-042f6d00-c913-11ea-8e8e-488e3eca8545.png)

with the debug mode enabled, nothing is showing in the logs.  
This is my compose setup:

```
  authelia:
    image: authelia/authelia:4.21
    container_name: authelia
    hostname: authelia 
    restart: unless-stopped
    volumes:
      - $USERDIR/authelia:/config
    networks:
      - reverseproxy
      - intern        
    environment:
      TZ: $TZ   
      AUTHELIA_JWT_SECRET_FILE: /run/secrets/authelia_jwt
      AUTHELIA_SESSION_SECRET_FILE: /run/secrets/authelia_session
      AUTHELIA_SESSION_REDIS_PASSWORD_FILE: /run/secrets/authelia_redis
      AUTHELIA_STORAGE_MYSQL_PASSWORD_FILE: /run/secrets/authelia_mariadb
      AUTHELIA_NOTIFIER_SMTP_PASSWORD_FILE: /run/secrets/smtp
      AUTHELIA_AUTHENTICATION_BACKEND_LDAP_PASSWORD_FILE: /run/secrets/ldap
    secrets:
      - authelia_mariadb
      - authelia_redis
      - authelia_jwt
      - authelia_session  
      - ldap
      - smtp
    depends_on:
      - authelia-mariadb
      - authelia-redis
      - openldap
      - openldap-admin
    expose:
      - 9091
    labels:
      traefik.enable: "true"
      ## HTTP Routers
      traefik.http.routers.authelia-rtr.entrypoints: https
      traefik.http.routers.authelia-rtr.rule: HostHeader(`login.$DOMAINNAME`)
      ## Middlewares
      traefik.http.routers.authelia-rtr.middlewares: chain-no-oauth@file           
      ## HTTP Services
      traefik.http.routers.authelia-rtr.service: authelia-svc
      traefik.http.services.authelia-svc.loadbalancer.server.port: 9091   
      ## watchtower       
      com.centurylinklabs.watchtower.enable: "true"        
```

What are the steps to get it working.

**Comment by clems4ever** [comment]
Hello @Akruidenberg , please check the logs in the browser console. Is there any more information there?
Can you please also paste the logs? I need to see what I can add if there is not enough logs.

Later, if we have not found any more clue, I'll also need the Authelia configuration too (redacted).

**Comment by Akruidenberg** [comment]
If I have some time, paste the info as soon as possible.

**Comment by wilhelminamontgomery** [comment]
Hi,

I have the same error message as @Akruidenberg. from portainer, the container log for authelia is as below which is something about token not in the database:



time="2020-10-08T06:06:30-04:00" level=info msg="Storage schema upgrade to v1 completed",
time="2020-10-08T06:06:30-04:00" level=info msg="Authelia is listening for non-TLS connections on 0.0.0.0:9091",
time="2020-10-08T06:07:29-04:00" level=error msg="Token is not in DB, it might have already been used" method=POST path=/api/secondfactor/totp/identity/finish remote_ip= stack="github.com/authelia/authelia/internal/middlewares/authelia_context.go:64       (*AutheliaCtx).Error\ngithub.com/authelia/authelia/internal/middlewares/identity_verification.go:142 IdentityVerificationFinish.func1\ngithub.com/authelia/authelia/internal/middlewares/require_first_factor.go:15   RequireFirstFactor.func1\ngithub.com/authelia/authelia/internal/middlewares/authelia_context.go:49       AutheliaMiddleware.func1.1\ngithub.com/fasthttp/router@v1.2.4/router.go:348                                (*Router).Handler\ngithub.com/authelia/authelia/internal/middlewares/log_request.go:14            LogRequestMiddleware.func1\ngithub.com/valyala/fasthttp@v1.15.1/server.go:2162                             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.15.1/workerpool.go:223                          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.15.1/workerpool.go:195                          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1374                                                       goexit",



do you need additional information from me?

**Comment by nightah** [comment]
@Akruidenberg did you manage to sort out this issue? If not can you please provide the aforementioned information?

@wilhelminamontgomery in your scenario it looks to me like the token has expired, given that the logs you have provided seem to be from a clean start up of Authelia is it possible that your token was generated much longer before starting up the container?

**Comment by nightah** [comment]
@Akruidenberg I'm closing this issue off, if you experience the issue again please provide the details requested and we can reopen this issue.

**Comment by ChaseDDevelopment** [comment]
Hey @nightah I ran into this issue tonight. This is my first time posting to a github issue, I hope I'm doing this right. I solved the problem but Thought I would post, and post my logs to maybe help see if this is a bigger issue or just something I did wrong. 

I had been setting up authelia and saving my notifications to the filesystem as I don't have smtp notifier. I mounted the /tmp/authelia when clicking on the link that is generated in the text file, I had this error consistently every time. 

I tried removing authelia container, obtaining a smtp notifier and even clearing the authelia database in mariadb. The final culprit was that the token generated with the notifier did not match what was being put into the _identity_verification_tokens_ table in the database. I modified the link by copying the token from the database and appending it to 

https://authelia.example.com/one-time-password/register?token=

and I was successful in generating the QR code.

Let me know if you need any additional information! I'm sure I can reproduce this in the future if you need more logs or anything. 

authelia container:

`
2021-01-02T12:03:40.712425664Z time="2021-01-02T05:03:40-07:00" level=error msg="Token is not in DB, it might have already been used" method=POST path=/api/secondfactor/totp/identity/finish remote_ip=#.#.#.# stack="github.com/authelia/authelia/internal/middlewares/authelia_context.go:64       (*AutheliaCtx).Error\ngithub.com/authelia/authelia/internal/middlewares/identity_verification.go:142 IdentityVerificationFinish.func1\ngithub.com/authelia/authelia/internal/middlewares/require_first_factor.go:15   RequireFirstFactor.func1\ngithub.com/authelia/authelia/internal/middlewares/authelia_context.go:49       AutheliaMiddleware.func1.1\ngithub.com/fasthttp/router@v1.2.4/router.go:348                                (*Router).Handler\ngithub.com/authelia/authelia/internal/middlewares/log_request.go:14            LogRequestMiddleware.func1\ngithub.com/valyala/fasthttp@v1.15.1/server.go:2162                             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.15.1/workerpool.go:223                          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.15.1/workerpool.go:195                          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1374      
`


Docker-Compose.yml

``` # Authelia (Lite) - Self-Hosted Single Sign-On and Two-Factor Authentication
  authelia:
    container_name: authelia
    image: authelia/authelia:latest
    restart: always
    networks:
      NETWORK:
        ipv4_address: #.#.#.#
    ports:
      - "$AUTHELIA_PORT:9091"
    depends_on:
      - mariadb
      - redis
    volumes:
      - $DOCKERDIR/authelia:/config
      - $DOCKERDIR/authelia/tmp:/tmp/authelia/ #This is just for grabbing notification. 
    environment:
      - TZ=$TZ
      - AUTHELIA_SESSION_REDIS_PASSWORD_FILE=/run/secrets/authelia_session_redis_password
      - AUTHELIA_JWT_SECRET_FILE=/run/secrets/authelia_jwt_secret
      - AUTHELIA_SESSION_SECRET_FILE=/run/secrets/authelia_session_secret
      - AUTHELIA_STORAGE_MYSQL_PASSWORD_FILE=/run/secrets/authelia_storage_mysql_password
    secrets:
      - authelia_jwt_secret
      - authelia_session_secret
      - authelia_storage_mysql_password
      - authelia_session_redis_password
    labels:
      - "traefik.enable=true"
      ## HTTP Routers
      - "traefik.http.routers.authelia-rtr.entrypoints=https"
      - "traefik.http.routers.authelia-rtr.rule=Host(`authelia.$DOMAINNAME`)"
      - "traefik.http.routers.authelia-rtr.tls=true"
      ## Middlewares
      - "traefik.http.routers.authelia-rtr.middlewares=chain-authelia@file"
      ## HTTP Services
      - "traefik.http.routers.authelia-rtr.service=authelia-svc"
      - "traefik.http.services.authelia-svc.loadbalancer.server.port=9091"
```
Authelia Config:

```
###############################################################
#                   Authelia configuration                    #
###############################################################

host: 0.0.0.0
port: 9091
log_level: warn

# https://docs.authelia.com/configuration/miscellaneous.html#default-redirection-url
default_redirection_url: https://authelia.example.com

totp:
  issuer: authelia.com
  period: 30
  skew: 1

authentication_backend:
  file:
    path: /config/users_database.yml
    # customize passwords based on https://docs.authelia.com/configuration/authentication/file.html
    password:
      algorithm: argon2id
      iterations: 1
      salt_length: 16
      parallelism: 8
      memory: 1024 # blocks this much of the RAM. Tune this.

# https://docs.authelia.com/configuration/access-control.html
access_control:
  default_policy: deny
  rules:
    # Rules applied to everyone
    - domain: authelia.example.com
      policy: bypass
    - domain: "*.example.com"
      policy: two_factor
    - domain: "example.com"
      policy: two_factor
session:
  name: authelia_session
  expiration: 3600 # 1 hour
  inactivity: 300 # 5 minutes
  domain: example.com

regulation:
  max_retries: 3
  find_time: 120
  ban_time: 300

storage:
  mysql:
    host: mariadb
    port: 3306
    database: {database}
    username: {database_user}

notifier:
  filesystem:
    filename: /tmp/authelia/notification.txt

```

**Comment by nightah** [comment]
@ChaseDDevelopment did you use the same browser/session to start the enrolment process and visit the link that was emailed?

I'll see if I can try replicate this with MariaDB, but I've definitely just tried with SQLite and not been able to do so.

The tokens that are generated are one time use and they need to be utilised by the same browser/session that initiated the process, utilising it by another browser will self destruct the token.

**Comment by ChaseDDevelopment** [comment]
@nightah I did use the same browser/session to start the process and finish the process. I am using firefox.

 The token remains in the database until it is used by the modified link in the browser that initiated the enrollment process. The notification.txt file is where the token is seemingly completely random. I've tried just adding one key at a time, so that there is only one key in the database, and the key that appears in the notification.txt file is completely different.

let me know if I can provide any more information. - as I said, I solved the issue, and if its not re-producable on your end, then it was probably just something whacky with my setup. 

Here is my maridb compose:

```
  #MariaDB - MySQL Database
  mariadb:
    container_name: mariadb
    image: mariadb:latest
    restart: always
    networks:
      NETWORK:
        ipv4_address: #.#.#.#
    security_opt:
      - no-new-privileges:true
    ports:
      - $MARIADBPORT:3306
    volumes:
      - $USERDIR/docker/mariadb/data:/config
      - /etc/timezone:/etc/timezone:ro
      - /etc/localtime:/etc/localtime:ro
      - mariadb:/var/lib/mysql
    deploy:
      resources:
        limits:
          cpus: '0.75'
          memory: 4096M
        reservations:
          memory: 1024M
    environment:
      - PUID=$PUID
      - PGID=$PGID
      - TZ=$TZ
      - MYSQL_DATABASE=$MYSQL_DATABASE
      - MYSQL_ROOT_PASSWORD=$MYSQL_ROOT_PASSWORD
      - MYSQL_USER=$MYSQL_USER
      - MYSQL_PASSWORD=$MYSQL_PASSWORD
      - REMOTE_SQL=$ABSOLUTE_URI

```

**Comment by nightah** [comment]
I've just tested with MariaDB as per the configuration you provided and am not able to reproduce this issue.

The link generated in the `notification.txt` along with the token works for me. The token also matches that which is in the DB.

Can I suggest you lodge another issue instead and include actual extracts of the links that are generated and what appears in your database please? Happy for you to redact sensitive information but it'd be good to see how the generated link differs from what is in the database.

**Comment by beaug** [comment]
I know this is old but just wanted to add I've the same issue (but later found it was a simple mistake). Error in Chrome browser as requested above is

Error: Failed POST to /api/secondfactor/totp/identity/finish. Code: 200. Message: The identity verification token has already been used
    at Client.ts:9
    at c (runtime.js:63)
    at Generator._invoke (runtime.js:294)
    at Generator.next (runtime.js:119)
    at r (asyncToGenerator.js:3)
    at l (asyncToGenerator.js:25)
(anonymous) @ RegisterOneTimePassword.tsx:49

Then in portainer log for authelia:
method GET) is not authorized to user <anonymous>, responding with status code 401  with location to redirect to [auth url]

As some of the others suggested above using the token from the DB seemed to make it work. 

Key from notification txt:

https://URL/one-time-password/register?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2MzMyMjQ4MDcsImlzcyI6IkF1dGhlbGlhIiwiYWN0aW9uIjoiUmVnaXN0ZXJUT1RQRGV2aWNlIiwidXNlcm5hbWUiOiJiZWF1Z2F1ZHJvbiJ9.x1Gvgdwqzow_zuRa4jBAmaoRF-nF3C7L-rWjAOEqMic

Key from database directly

https://URL/one-time-password/register?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2MzMyMjQ4MDcsImlzcyI6IkF1dGhlbGlhIiwiYWN0aW9uIjoiUmVnaXN0ZXJUT1RQRGV2aWNlIiwidXNlcm5hbWUiOiJiZWF1Z2F1ZHJvbiJ9.x1Gvgdwqzow_zuRa4jBAmaoRF-nF3C7L-rWjAOEqMic

I copied the key from the DB directly and the QR code generated fine, which made me think maybe there is an issue between the notification generated and what gets stored in the DB. I used some tools to check the keys, they are identical so no issues there. 

So I then tried to replicate the problem - turns out in my case because I was opening the notification file in nano over SSH part of the key was actually not showing and I just assumed that was the end of the line as it looked like it was. 

How did I know? Use the command tail notification.txt instead of opening the file in an editor as a terminal will just carry the URL over two lines. 

Hopefully this helps and its this simple for someone else
